Data/Row/Internal.hs

```diff
@@ -282,42 +275,39 @@ instance (KnownSymbol ℓ, c τ, FoldStep ℓ τ ρ, Forall ('R ρ) c) => Forall
                -- ^ The way to transform the empty element
             -> (forall ℓ τ ρ. (KnownSymbol ℓ, c τ) => Label ℓ -> f ('R (ℓ :-> τ ': ρ)) -> Either (h τ) (f ('R ρ)))
                -- ^ The unfold
-            -> (forall ℓ τ ρ. (KnownSymbol ℓ, c τ, FoldStep ℓ τ ρ) => Label ℓ -> Either (h τ) (g ('R ρ)) -> g ('R (ℓ :-> τ ': ρ)))
+            -> (forall ℓ τ ρ. (KnownSymbol ℓ, c τ) => Label ℓ -> Either (h τ) (g ('R ρ)) -> g ('R (ℓ :-> τ ': ρ)))
                -- ^ The fold
             -> f ('R (ℓ :-> τ ': ρ))  -- ^ The input structure
             -> g ('R (ℓ :-> τ ': ρ))
   {-# INLINE metamorph' #-}
   metamorph' _ empty uncons cons r = cons Label $ metamorph' @_ @('R ρ) @c @_ @_ @h Proxy empty uncons cons <$> uncons Label r
 
--- | Any structure over two rows in which every element of both rows satisfies the
---   given constraint can be metamorphized into another structure over both of the
+-- | Any structure over two rows in which the elements of each row satisfy some
+--   constraints can be metamorphized into another structure over both of the
 --   rows.
--- TODO: Perhaps it should be over two constraints?  But this hasn't seemed necessary
---  in practice.
-class Forall2 (r1 :: Row k) (r2 :: Row k) (c :: k -> Constraint) where
-  -- | A metamorphism is a fold followed by an unfold.  Here, we fold both of the inputs.
-  metamorph2 :: forall (f :: Row k -> *) (g :: Row k -> *) (h :: Row k -> Row k -> *)
-                       (f' :: k -> *) (g' :: k -> *).
-                Proxy f' -> Proxy g'
-             -> (f Empty -> g Empty -> h Empty Empty)
-             -> (forall ℓ τ1 τ2 ρ1 ρ2. (KnownSymbol ℓ, c τ1, c τ2)
-                 => Label ℓ
-                 -> f ('R (ℓ :-> τ1 ': ρ1))
-                 -> g ('R (ℓ :-> τ2 ': ρ2))
-                 -> ((f' τ1, f ('R ρ1)), (g' τ2, g ('R ρ2))))
-             -> (forall ℓ τ1 τ2 ρ1 ρ2. (KnownSymbol ℓ, c τ1, c τ2)
-                 => Label ℓ -> f' τ1 -> g' τ2 -> h ('R ρ1) ('R ρ2) -> h ('R (ℓ :-> τ1 ': ρ1)) ('R (ℓ :-> τ2 ': ρ2)))
-             -> f r1 -> g r2 -> h r1 r2
-
-instance Forall2 (R '[]) (R '[]) c where
-  {-# INLINE metamorph2 #-}
-  metamorph2 _ _ empty _ _ = empty
-
-instance (KnownSymbol ℓ, c τ1, c τ2, Forall2 ('R ρ1) ('R ρ2) c)
-      => Forall2 ('R (ℓ :-> τ1 ': ρ1)) ('R (ℓ :-> τ2 ': ρ2)) c where
-  {-# INLINE metamorph2 #-}
-  metamorph2 f g empty uncons cons r1 r2 = cons (Label @ℓ) t1 t2 $ metamorph2 @_ @('R ρ1) @('R ρ2) @c f g empty uncons cons r1' r2'
-    where ((t1, r1'), (t2, r2')) = uncons (Label @ℓ) r1 r2
+class BiForall (r1 :: Row k1) (r2 :: Row k2) (c1 :: k1 -> Constraint) (c2 :: k2 -> Constraint) where
```

Comment:
Why not `BiForall (r1 :: Row k1) (r2 :: Row k2) (c :: k1 -> k2 -> Constraint)`? It would be more general.

Reply:
It is more general, and it was the first iteration of this work, but it's super painful to use.  For instance, if `r1` contains functors and `r2` contains arbitrary values, then it's easy to say `c1` is `Functor` and `c2` is `Unconstrained1`, but I know of no easy way to describe a two-argument `c` that captures this.  A constraint synonym would be nice, but those must be fully applies, so the only option is a class.  For my examples, I ended up writing:
```haskell
class Functor f => FirstFunctor f x
instance Functor f => FirstFunctor f x

class (Functor f, Show x) => FunctorAndShow f x
instance (Functor f, Show x) => FunctorAndShow f x
```
before deciding that having two constraints was much easier.